pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionMetaDataManager.java

```diff
@@ -168,68 +151,78 @@ public synchronized boolean containsFunction(String tenant, String namespace, St
+        if (tailer != null) {
+            tailer.stopWhenNoMoreMessages();
+            tailer.close();
+        }
+        this.schedulerManager.schedule();
+        log.info("FunctionMetaDataManager done becoming leader by doing its first schedule");
     }
 
-    /**
-     * Sends a start/stop function request to the FMT (Function Metadata Topic) for a function
-     * @param tenant the tenant the function that needs to be deregistered belongs to
-     * @param namespace the namespace the function that needs to be deregistered belongs to
-     * @param functionName the name of the function
-     * @param instanceId the instanceId of the function, -1 if for all instances
-     * @param start do we need to start or stop
-     * @return a completable future of when the start/stop has been applied
-     */
-    public synchronized CompletableFuture<RequestResult> changeFunctionInstanceStatus(String tenant, String namespace, String functionName,
-                                                                                      Integer instanceId, boolean start) {
-        FunctionMetaData functionMetaData = this.functionMetaDataMap.get(tenant).get(namespace).get(functionName);
-
-        FunctionMetaData newFunctionMetaData = FunctionMetaDataUtils.changeFunctionInstanceStatus(functionMetaData, instanceId, start);
-
-        Request.ServiceRequest updateRequest = ServiceRequestUtils.getUpdateRequest(
-                this.workerConfig.getWorkerId(), newFunctionMetaData);
+    private synchronized FunctionMetaDataTopicTailer internalAcquireLeadership() {
+        if (exclusiveLeaderProducer == null) {
+            try {
+                exclusiveLeaderProducer = pulsarClient.newProducer()
+                        .topic(this.workerConfig.getFunctionMetadataTopic())
+                        .producerName(workerConfig.getWorkerId() + "-leader")
+                        // .type(EXCLUSIVE)
+                        .create();
+            } catch (PulsarClientException e) {
+                log.error("Error creating exclusive producer", e);
+                errorNotifier.triggerError(e);
+            }
+        } else {
+            log.error("Logic Error in FunctionMetaData Manager");
+            errorNotifier.triggerError(new IllegalStateException());
+        }
+        FunctionMetaDataTopicTailer tailer = this.functionMetaDataTopicTailer;
+        this.functionMetaDataTopicTailer = null;
+        return tailer;
+    }
 
-        return submit(updateRequest);
+    public synchronized void giveupLeadership() {
+        log.info("FunctionMetaDataManager giving up leadership by closing exclusive producer");
+        try {
+            exclusiveLeaderProducer.close();
+        } catch (PulsarClientException e) {
+            log.error("Error closing exclusive producer", e);
+            errorNotifier.triggerError(e);
+        }
+        exclusiveLeaderProducer = null;
+        initializeTailer();
```

Comment:
I is not a good idea to simply re-create the metadata tailer object.  This will cause the tailer to start reading from the beginning.  This will cause all the tailer to have to re-read the whole topic.  The problem here is not only that the the tailer has to re-read the whole topic which can take time but during that time the in-memory metadata map for the worker will be inconsistent.  If a user request is sent to the worker during this time, the behavior might be incorrect.  I would suggest keeping track of the message Id that represents to which message the current view of the metadata relates to.  When the worker is the leader and it updates its in memory metadata cache and  produces messages to metadata topic, we should update the message id.  When the worker loses leadership, the tailer should use the message id and start reading from there

Reply:
will do

Reply:
changed